Flake8 plugin for linting

youben11 opened the pull request:
As per the discussion here #190 a way of supporting flake8 linting is by implementing an additional plugin that will use the flake8 config directly instead of passing it to other linting tools like pycodestyle and pyflakes (even if flake8 is a wrapper around those tools by itself).

Flake8 lacks (for now) of a stable [public python API](http://flake8.pycqa.org/en/latest/user/python-api.html) which should be implemented in the future, so I figured out a dirty way from the source code to access the report results and get the information we need to report back to the IDE, however, this should be updated whenever a python API is made available.

youben11 committed:
new plugin flake8 for linting

youben11 commented:
@gatesn

uschi2000 commented on pyls/plugins/flake8_lint.py:
```diff
@@ -0,0 +1,61 @@
+# Copyright 2017 Palantir Technologies, Inc.
```
2019

uschi2000 commented on pyls/plugins/flake8_lint.py:
```diff
@@ -0,0 +1,61 @@
+
+log = logging.getLogger(__name__)
+
+
```
can you add some dev docs, please: what is this, what does it do? please also think about user-facing documentation (@gatesn , where do we keep those?)

youben11 replied:
of course I will

uschi2000 commented on pyls/plugins/flake8_lint.py:
```diff
@@ -0,0 +1,61 @@
+    report = style_guide.check_files([document.path])
+    diagnostics = parse_report(document, report)
+
+    return diagnostics
```
inline return statement

uschi2000 commented on pyls/plugins/flake8_lint.py:
```diff
@@ -0,0 +1,61 @@
+    opts = {
+        'exclude': settings.get('exclude'),
+        'filename': settings.get('filename'),
+        'hang_closing': settings.get('hangClosing'),
```
what's the deal with camelCase vs snake_case here?

youben11 replied:
I got that part from another plugin using the same conf, but it looks like it wasn't working, so I'm willing to debug this to fix it

gatesn replied:
VSCode settings convention appears to be camel case, but we need to pass it to the Flake8 as a snake case option

youben11 replied:
@gatesn how does it affects VSCode if for example we rename plugins.pycodestyle.hangClosing to plugins.pycodestyle.hang-closing ?

gatesn replied:
Renaming configs would likely break existing users. We could build utility functions to convert between case styles, but not sure it's that bad as-is

youben11 replied:
got it

uschi2000 commented on setup.py:
```diff
@@ -76,6 +76,7 @@
         ],
         'pyls': [
             'autopep8 = pyls.plugins.autopep8_format',
+            'flake8 = pyls.plugins.flake8_lint',
```
possible to add a test for this functionality?

youben11 replied:
done

uschi2000 commented on pyls/plugins/flake8_lint.py:
```diff
@@ -0,0 +1,61 @@
+        log.error("Flake8 parsed more than a file for '%s'", document.path)
+
+    diagnostics = []
+    file_checker = file_checkers[0]
```
do you need to guard against `len(file_checkers) == 0`, or are you guaranteed that there is one element at least?

youben11 replied:
filecheckers is a list of checkers for files that are being linted, since the pyls_lint hook should receive only one file to lint, having more than one is incorrect, I should also check for `len(file_checkers) == 0` to return an empty list also

uschi2000 commented on pyls/plugins/flake8_lint.py:
```diff
@@ -0,0 +1,61 @@
+    file_checkers = report._application.file_checker_manager.checkers
+    # There should be only one filechecker since we are parsing using a path and not a pattern
+    if len(file_checkers) > 1:
+        log.error("Flake8 parsed more than a file for '%s'", document.path)
```
s/a file/one file/

youben11 committed:
typos

youben11 committed:
calling func in retrun statement

youben11 committed:
Dealing with the case where no file is checked

youben11 committed:
flake8 dev docs

youben11 committed:
adding flake8 to extra packages.

The version is specified due to the hard way the code uses flake8
that might not work if a code change was introduced in flake8 codebase

youben11 committed:
testing flake8 plugin

youben11 committed:
typo

youben11 commented:
Test is failing because of the latest release of pylint  #661

uschi2000 commented:
Looks good to me!

youben11 committed:
Merge branch 'develop' into develop

youben11 commented:
@gatesn As flake8_config is now aggregating flake8 config for both pycodestyle and mccabe, we should update it to load the config under plugins.falke8, right? Users will no longer be able to use flake8 config files with pycodestyle and mccabe that way, the flake8 plugin will be the only one to use flake8 config files.

gatesn commented on pyls/plugins/flake8_lint.py:
```diff
@@ -0,0 +1,93 @@
+    # No file have been checked
+    if not file_checkers:
+        return []
+    # There should be only a filechecker since we are parsing using a path and not a pattern
```
We can collapse these two statements into `if len(blah) != 1`

youben11 replied:
I don't see the way we can do that because they are totally different. The first reflects the case where a file being linted haven't been saved yet so we should return no errors, the second might happen if the filename was interpreted as a pattern, I did this check cause the internal API may receive both filename and pattern so we must log the case where it's not working as expected.

gatesn commented on setup.py:
```diff
@@ -48,6 +48,7 @@
     extras_require={
         'all': [
             'autopep8',
+            'flake8==3.7.8',
```
Even though we use internal APIs, let’s not pin this. Otherwise we’ll end up conflicting with people’s own Flake8 versions.

We should just track latest the best we can I reckon

youben11 replied:
done

gatesn commented on pyls/plugins/flake8_lint.py:
```diff
@@ -0,0 +1,93 @@
+
+
+@hookimpl
+def pyls_lint(config, document):
```
Should we default this plugin to disabled, and when it is enabled disable pep8 and pyflakes? I think at the moment we’d end up with duplicate diagnostics which could be annoying

youben11 replied:
yeah, they shouldn't be running at the same time, and disabling flake8 by default is a good idea so current users don't feel the change but can enable linting directly via flake8 if they want to. But flake8_config have to be made loading config into flake8 plugin as well from now on.

youben11 committed:
don't pinpoint flake8 version

youben11 committed:
default flake8 plugin to disabled

youben11 committed:
flake8 plugin config

youben11 committed:
Merge branch 'develop' of github.com:youben11/python-language-server into develop

gatesn committed:
Merge branch 'develop' into develop
